Add custom 404 page with Neovim branding

alarcritty opened the pull request:
## Summary
Adds a custom 404 page that matches the Neovim website design and provides helpful navigation for users who encounter broken links.

fix #382

## Problem
Currently, non-existent URLs on neovim.io show GitHub's default 404 page, which doesn't match the site's branding and doesn't help users navigate back to relevant content.

## Solution
- Created `404.html` using Jekyll's layout system with `{% include nav.html %}`
- Matches the existing Neovim website color scheme and typography
- Includes working navigation header with search functionality
- Provides helpful links to popular pages (Documentation, News, GitHub, etc.)
- Responsive design that works on mobile and desktop
- Includes auto-redirect logic for common typos (docs → doc/, etc.)
- Tracks 404 errors for analytics if available

## Features
- ✅ Uses existing Jekyll layout and includes for consistency
- ✅ Matches Neovim brand colors and styling
- ✅ Responsive design
- ✅ Working navigation header
- ✅ Helpful links to popular pages
- ✅ Auto-redirect for common URL mistakes
- ✅ Analytics tracking support

## Testing
- [x] Tested locally with `jekyll serve` 
- [x] 404 page displays correctly at `/404`
- [x] Uses existing site CSS variables and components
- [x] Mobile responsive design verified

justinmk commented on 404.html:
```diff
@@ -0,0 +1,152 @@
+{% include nav.html %}
+<main class="front-section">
+  <div class="container">
+    <div class="error-content">
+      <div class="error-code">404</div>
+      <h1>Page Not Found</h1>
```
why is `error-content` needed? why can't we use the standard section class names. also not clear why 404 is in its own div
```suggestion
    <div class="...">
      <h1>404 Page Not Found</h1>
```

alarcritty replied:
ah yes, you're right. switched to using .masthead instead and put the 404 in .lead.

justinmk commented on 404.html:
```diff
@@ -0,0 +1,152 @@
+  flex-wrap: wrap;
+}
+
+.helpful-links {
```
why do we need classes defined for this

the entire purpose of CSS is to re-use classes, i don't get why people always define new classes for every single thing

alarcritty replied:
totally agree, was overthinking it. killed that whole section.

justinmk commented on 404.html:
```diff
@@ -0,0 +1,152 @@
+        <a href="/doc/" class="btn">Documentation</a>
+      </div>
+      
+      <div class="helpful-links">
+        <h3>Popular Pages</h3>
+        <div class="col3">
+          <a href="/" class="link-item">Home</a>
+          <a href="/news/" class="link-item">News</a>
+          <a href="/doc/" class="link-item">Documentation</a>
+          <a href="/sponsors/" class="link-item">Sponsors</a>
+          <a href="/charter/" class="link-item">Charter</a>
+          <a href="https://github.com/neovim/neovim" class="link-item">GitHub</a>
```
we can drop this, it's redundant with the nav, and the 2 buttons are good enough.

alarcritty replied:
yes, removed it entirely. nav + buttons is plenty.

alarcritty commented on 404.html:
```diff
@@ -0,0 +1,152 @@
+      <h1>Page Not Found</h1>
+      <p class="light">The page you're looking for doesn't exist. It might have been moved, deleted, or you entered the wrong URL.</p>
+      
+      <div class="flex error-buttons">
```
oh okay so i , dropped the error-buttons class. just .flex + .btn now, same as homepage.

alarcritty commented:
fixed all of these:
- using .masthead instead of custom wrapper
- 404 uses existing .lead class 
- killed the redundant links section
- removed error-buttons, just .flex now
- basically eliminated all the custom classes

also this is how page looks now .
<img width="1675" height="1013" alt="2025-09-14_10-44" src="https://github.com/user-attachments/assets/ac6038b7-5805-4f54-881a-c6c008c71322" />

should be much cleaner now, let me know if anything else looks off

justinmk commented:
The color and typewriter style of "404" in https://github.com/neovim/neovim.github.io/issues/382#issuecomment-3284133829 looks better. If that requires a custom css then ok.

justinmk commented on 404.html:
```diff
@@ -0,0 +1,93 @@
+    <div class="masthead">
+      <div class="lead">404</div>
+      <h1>Page Not Found</h1>
+      <p class="light">The page you're looking for doesn't exist. It might have been moved, deleted, or you entered the wrong URL.</p>
```
Restricting this to the middle "third" as in https://github.com/neovim/neovim.github.io/issues/382#issuecomment-3284133829 looked better

alarcritty replied:
want me to use the existing golden-grid class to constrain it to ~60% width, or just add a max-width on the masthead?

justinmk replied:
> want me to use the existing golden-grid class to constrain it to ~60% width

yes

justinmk commented on 404.html:
```diff
@@ -0,0 +1,93 @@
+</style>
+
+<script>
+// Track 404 errors 
+if (typeof gtag !== 'undefined') {
+  gtag('event', 'page_view', {
+    page_title: '404 - Page Not Found',
+    page_location: window.location.href
+  });
+}
```
??

alarcritty commented on 404.html:
```diff
@@ -0,0 +1,93 @@
+  });
+}
+
+// Auto-redirect after 3s
+const currentPath = window.location.pathname.toLowerCase();
+const redirects = {
+  '/docs': '/doc/',
+  '/documentation': '/doc/',
+  '/github': 'https://github.com/neovim/neovim',
+  '/repo': 'https://github.com/neovim/neovim'
+};
+
+if (redirects[currentPath]) {
+  console.log(`Redirecting ${currentPath} to ${redirects[currentPath]} in 3 seconds...`);
+  setTimeout(() => {
+    window.location.href = redirects[currentPath];
+  }, 3000);
+}
```
this auto-redirects common typos like /docs → /doc/ after 3 seconds. helps users who mistype URLs get to the right place automatically.

justinmk commented:
one last nit then lgtm!

alarcritty commented on 404.html:
```diff
@@ -0,0 +1,71 @@
+  color: #61ff00;
+  font-family: 'SF Mono', 'Monaco', 'Inconsolata', 'Roboto Mono', 'Courier New', monospace;
+  line-height: 1;
+  margin-bottom: 1rem;
```
sure, what margin would work better? 0.5rem or even less?

alarcritty commented:
also this is how it looks now after refactoring to golden grid 
<img width="1366" height="721" alt="2025-09-15_01-10" src="https://github.com/user-attachments/assets/24dc6741-267a-4220-8ad3-11bfa3382b6e" />

alarcritty commented:
> can the "404" have the color from [#382 (comment)](https://github.com/neovim/neovim.github.io/issues/382#issuecomment-3284133829) instead of the neon green

done , here you go 
<img width="1366" height="727" alt="2025-09-15_05-36" src="https://github.com/user-attachments/assets/ba85ca72-70cb-4648-b4e9-b7632908fd47" />
also, let me know if there is anything else to be done .

alarcritty commented:
> Let's try it, thanks!

Many thanks for the review and merge. I’m eager to keep contributing further.

alarcritty commented:
hey @justinmk ,
also i see one issue there , when i do /non-existed page , it still shows me the old github 404 page , why's that

alarcritty commented:
okay there is a slight change , i just need to include this line in _layouts/default.html file ,
      <link rel="canonical" href="{{ site.url }}{% if page.canonical_url %}{{ page.canonical_url }}{% else %}{{ page.url }}{% endif %}" />
, so should i open another pr ?

alarcritty commented:
{% unless page.url == "/404.html" %}
  <link rel="canonical" href="{{ site.url }}{% if page.canonical_url %}{{ page.canonical_url }}{% else %}{{ page.url }}{% endif %}" />
{% else %}
  <meta name="robots" content="noindex">
{% endunless %}

also we need to update the _layouts/default.html , so should i open the new pr ?

justinmk commented:
> this is the change we need to do in _layouts/default.html file , so should i open another pr

yes, thanks
